File: AlgorithmsProject/src/in/algorithms/tree/BalancedTreeChecker.scala
package in.algorithms.tree

class BalancedTreeChecker {

  def minDepth(treeNode: TreeNode): Integer = {
    if (treeNode == null)
      0
    else 1 + Math.min(minDepth(treeNode.left), minDepth(treeNode.right))
  }

  def maxDepth(treeNode: TreeNode): Integer = {
    if (treeNode == null)
      0
    else 1 + Math.max(maxDepth(treeNode.left), maxDepth(treeNode.right))
  }

  def isBalanced(treeNode: TreeNode): Boolean = {
    maxDepth(treeNode) - minDepth(treeNode) <= 1
  }

}

File: AlgorithmsProject/src/in/algorithms/tree/BalancedTreeChecker.java
package in.algorithms.tree;

public class BalancedTreeChecker {
    public int minDepth(TreeNode treeNode) {
        if (treeNode == null) return 0;
        return 1 + Math.min(minDepth(treeNode.left), minDepth(treeNode.right));
    }

    public int maxDepth(TreeNode treeNode) {
        if (treeNode == null) return 0;
        return 1 + Math.max(maxDepth(treeNode.left), maxDepth(treeNode.right));
    }

    public boolean isBalanced(TreeNode treeNode) {
        return Math.abs(maxDepth(treeNode) - minDepth(treeNode)) <= 1;
    }
}
